File: backend/src/main/java/com/smartconstat/model/Constat.java
package com.smartconstat.model;

import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "constats")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class Constat {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    private String lieu;
    private LocalDateTime dateTime;

    // Statut du constat (Non examiné, En cours d'exécution, Traité, Rejeté)
    @Builder.Default
    private String statut = "Non examiné";

    @Builder.Default
    private String priorite = "Normale"; // "Basse", "Normale", "Haute", "Urgente"

    // Véhicule A
    private String assureurA;
    private String contratA;
    private String nomA;
    private String prenomA;
    private String adresseA;
    private String vehiculeMarqueA;
    private String vehiculeModeleA;
    private String immatriculationA;
    private String paysA;
    private String sensSuiviA;

    // Véhicule B
    private String assureurB;
    private String contratB;
    private String nomB;
    private String prenomB;
    private String adresseB;
    private String vehiculeMarqueB;
    private String vehiculeModeleB;
    private String immatriculationB;
    private String paysB;
    private String sensSuiviB;

    // Dégâts
    private String pointChocInitial;

    @Column(columnDefinition = "TEXT")
    private String degatsApparentsA;

    @Column(columnDefinition = "TEXT")
    private String degatsApparentsB;

    @Column(columnDefinition = "TEXT")
    private String autresDegats;

    @Column(columnDefinition = "TEXT")
    private String circonstances; // stored as comma-separated

    @Column(columnDefinition = "TEXT")
    private String observations;

    private String temoins;
    @Builder.Default
    private boolean blesses = false;
    @Builder.Default
    private boolean degatsMaterielsAutres = false;
    @Builder.Default
    private boolean interventionPolice = false;

    private String croquisPath;
    private String signatureAPath;
    private String signatureBPath;
    
    @Column(columnDefinition = "TEXT")
    private String photosPaths;

    // Employé qui traite ce constat
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "traite_par_id")
    private User traitePar;

    private String traiteParNom;

    private LocalDateTime dateLimite;

    @Column(columnDefinition = "TEXT")
    private String noteInterne;

    @Column(columnDefinition = "TEXT")
    private String documentsManquants;

    @Column(columnDefinition = "TEXT")
    private String motifRejet;

    @Column(columnDefinition = "TEXT")
    private String commentaireDecision;

    private String responsabiliteEstimee;
    private Double montantEstime;
    @Builder.Default
    private boolean escalade = false;
    private String escaladeRaison;
    @Builder.Default
    private String factureStatut = "Non demandee";
    @Builder.Default
    private Double montantFacturesTotal = 0.0;
    @Builder.Default
    private Integer nombreFactures = 0;

    @Column(columnDefinition = "TEXT")
    private String priseEnChargeDecision;

    private LocalDateTime updatedAt;

    @Column(updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        if (statut == null) statut = "Non examiné";
        if (priorite == null) priorite = "Normale";
        if (factureStatut == null) factureStatut = "Non demandee";
        if (montantFacturesTotal == null) montantFacturesTotal = 0.0;
        if (nombreFactures == null) nombreFactures = 0;
        if (dateLimite == null) dateLimite = createdAt.plusDays(3);
        updatedAt = createdAt;
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: frontend_angular/src/app/core/services/constat.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Constat {
  id?: number;
  accidentId?: string;
  dateTime?: string;
  lieu?: string;
  statut?: string;
  userName?: string;
  userEmail?: string;
  traiteParNom?: string;
  traiteParId?: number;
  priorite?: string;
  dateLimite?: string;
  noteInterne?: string;
  documentsManquants?: string;
  motifRejet?: string;
  commentaireDecision?: string;
  responsabiliteEstimee?: string;
  montantEstime?: number;
  factureStatut?: string;
  montantFacturesTotal?: number;
  nombreFactures?: number;
  priseEnChargeDecision?: string;
  factures?: Array<{
    id?: number;
    mois?: string;
    montant?: number;
    echeance?: string;
    statut?: string;
    typeFacture?: string;
    photoUrl?: string;
    decisionStatut?: string;
    decisionCommentaire?: string;
    decisionAt?: string;
  }>;
  escalade?: boolean;
  escaladeRaison?: string;
  updatedAt?: string;
  
  // Véhicule A
  nomA?: string;
  prenomA?: string;
  assureurA?: string;
  contratA?: string;
  immatriculationA?: string;
  vehiculeMarqueA?: string;
  vehiculeModeleA?: string;
  adresseA?: string;
  degatsApparentsA?: string;
  
  // Véhicule B
  nomB?: string;
  prenomB?: string;
  assureurB?: string;
  contratB?: string;
  immatriculationB?: string;
  vehiculeMarqueB?: string;
  vehiculeModeleB?: string;
  adresseB?: string;
  degatsApparentsB?: string;
  
  // Détails
  circonstances?: string[];
  observations?: string;
  temoins?: string;
  pointChocInitial?: string;
  autresDegats?: string;
  blesses?: boolean;
  interventionPolice?: boolean;
  
  // Fichiers uploadés
  croquisPath?: string;
  signatureAPath?: string;
  signatureBPath?: string;
  photosPaths?: string;
}

@Injectable({
  providedIn: 'root'
})
export class ConstatService {
  // Using the local IP since this is a mobile environment test backend
  private apiUrl = 'http://192.168.1.189:8082/api/constats';

  constructor(private http: HttpClient) { }

  getAllConstats(statut?: string): Observable<Constat[]> {
    let url = `${this.apiUrl}/all`;
    if (statut) {
      url += `?statut=${encodeURIComponent(statut)}`;
    }
    return this.http.get<Constat[]>(url);
  }

  updateStatut(id: number, statut: string, comment?: string): Observable<any> {
    return this.http.put(`${this.apiUrl}/${id}/statut`, { statut, comment });
  }

  assignConstat(id: number, employeId: number, comment?: string): Observable<any> {
    return this.http.put(`${this.apiUrl}/${id}/assign`, { employeId, comment });
  }

  updateTracking(id: number, payload: any): Observable<any> {
    return this.http.put(`${this.apiUrl}/${id}/tracking`, payload);
  }

  getHistory(id: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/${id}/history`);
  }

  getAdminNotifications(): Observable<any> {
    return this.http.get(`${this.apiUrl}/admin/notifications`);
  }

  getEmployeePerformance(): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/admin/performance`);
  }
}
